sematic/tests/test_graph.py

```diff
@@ -0,0 +1,278 @@
+
+    assert output == 17
+
+    runs, artifacts, edges = api_client.get_graph(future.id, root=True)
+
+    runs_by_id = {run.id: run for run in runs}
+
+    graph = Graph(runs=runs, edges=edges, artifacts=artifacts)
+
+    (
+        futures_by_original_id,
+        input_artifacts,
+        output_artifacts,
+    ) = graph.clone_futures(storage=mock_local_resolver_storage)
+
+    assert len(futures_by_original_id) == len(runs)
+
+    root_future = futures_by_original_id[future.id]
+    assert root_future.calculator._func is pipeline._func
+    assert root_future.parent_future is None
+
+    for original_run_id, future_ in futures_by_original_id.items():
+        original_run = runs_by_id[original_run_id]
+        # No reset so state should be the same
+        assert future.state.value == original_run.future_state
+
+        # Parent future should exist and be the correct one
+        if original_run.parent_id is not None:
+            parent_run = runs_by_id[original_run.parent_id]
+            assert futures_by_original_id[parent_run.id] is future_.parent_future
+
+        # Making sure all future kwargs are correct
+        for input_edge in graph.edges_by_destination_id[original_run.id]:
+            artifact = graph.artifacts_by_id[input_edge.artifact_id]
+            assert input_artifacts[future_.id][input_edge.destination_name] is artifact
+
+            if input_edge.source_run_id is None:
+                value = get_artifact_value(artifact, mock_local_resolver_storage)
+                assert future_.kwargs[input_edge.destination_name] == value
+            else:
+                upstream_future = futures_by_original_id[input_edge.source_run_id]
+                assert future_.kwargs[input_edge.destination_name] is upstream_future
+
+        # Making sure output values are correct
+        for output_edge in graph.edges_by_source_id[original_run_id]:
+            artifact = graph.artifacts_by_id[output_edge.artifact_id]
+            assert output_artifacts[future_.id] is artifact
+
+            value = get_artifact_value(artifact, mock_local_resolver_storage)
+            assert future_.value == value
+            if output_edge.destination_run_id is not None:
+                downstream_run = runs_by_id[output_edge.destination_run_id]
+                downstream_future = futures_by_original_id[downstream_run.id]
+                assert downstream_future.kwargs[output_edge.destination_name] is future_
+
+
+def test_clone_futures_reset(
+    mock_auth,  # noqa: F811
+    mock_socketio,  # noqa: F811
+    test_db,  # noqa: F811
+    mock_local_resolver_storage,  # noqa: F811
+    mock_requests,  # noqa: F811
+):
+    future = pipeline(1, 2, 3)
+    resolver = LocalResolver()
+    future.resolve(resolver)
+
+    runs, artifacts, edges = api_client.get_graph(future.id, root=True)
+
+    graph = Graph(runs=runs, edges=edges, artifacts=artifacts)
+
+    reset_from_run_id = future.nested_future.kwargs["a"].nested_future.kwargs["a"].id
```

Comment:
😎 I'll have to remember this approach next time I want a run that's in the middle of the DAG somewhere.